src/librustc_mir/interpret/operand.rs

```diff
@@ -598,10 +598,18 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 throw_inval!(TooGeneric),
             ty::ConstKind::Unevaluated(def_id, substs) => {
                 let instance = self.resolve(def_id, substs)?;
-                return Ok(OpTy::from(self.const_eval_raw(GlobalId {
+                let param_env = if self.tcx.is_static(def_id) {
+                    ty::ParamEnv::reveal_all()
+                } else {
+                    self.param_env
+                };
+                let val = self.tcx.const_eval(param_env.and(GlobalId {
```

Comment:
Also please explain why we use `const_eval` here. There are still a few calls to `const_eval_raw` in the `interpret` folder, so why the inconsistency?

Reply:
This also makes me wonder why we have the `const_eval_raw` query at all if we do not use it here -- as far as I am concerned, the only reason that query exists is for its use in Miri?

Reply:
I think we still need them right now, but we can start getting rid of it once we have fully eliminated `PlaceBase::Static`. This is a cleanup that's slowly happening, so some inconsistency is to be expected I think.

Reply:
I am not convinced we need to be inconsistent as an intermediate state -- we could just wait until we can switch them all over, if that's really what we want. I am worried the cleanup will go half-way and then stall, or something like that.

But I also clearly lost the overview for what the heck is going on in the big picture, so whatever.^^ I just long for the days when I actually thought I understood the way the core Miri engine interacted with the rest of the compiler. But those days are gone.

Reply:
I added a comment explaining it. I don't even think it's too inconsistent considering that this is `operand.rs`, so we *read* values, while the other uses are just referencing things

Reply:
The other uses are creating things from which one can read as well, though. So this by no means guards *all* reads.

I guess this will guard all consts, as those are operands, but not affects statics or promoteds as those are not?

Reply:
> I added a comment explaining i

I saw the comment explaining the `ParamEnv`, but nothing for `_raw` vs not-raw?

Reply:
oops, should have pushed

Reply:
> The other uses are creating things from which one can read as well, though. So this by no means guards all reads.

yea, this is the inconsistent part.

> I guess this will guard all consts, as those are operands, but not affects statics or promoteds as those are not?

After #67000 all promoteds will be constants and all statics are already evaluated constants of the form `ConstValue::Scalar(Scalar::Ptr(indirect_static_alloc_id))`. Thus evaluating any unevaluated constant can safely use `const_eval`. Fetching a static's allocation will keep using `const_eval_raw`, because `const_eval` would trigger validation and if that static contains a relocation into the currently interpreted static, then we'd get a cycle.

Reply:
> Fetching a static's allocation will keep using const_eval_raw, because const_eval would trigger validation and if that static contains a relocation into the currently interpreted static, then we'd get a cycle.

This is still inconsistent though as it means eventually also reading from that static will not be validated.
Wouldn't the recursion prevention in validation prevent fetching the allocation recursively?

Reply:
> Wouldn't the recursion prevention in validation prevent fetching the allocation recursively?

The problem is that the query system errors out *before* we hit the allocation recursion prevention